service/tracker/execution.go

```diff
@@ -171,6 +189,48 @@ func (e *Execution) processNext() error {
 	return nil
 }
 
+func (e *Execution) checkTrieUpdates(record *uploader.BlockData, blockID flow.Identifier) error {
+	e.log.Debug().Hex("block_id", blockID[:]).Msg("fetching updates from data sync")
+	e.log.Debug().Int("updates", len(record.TrieUpdates)).Msg("got trie updates from GCP")
+	// get Trie updates from exec data sync
+	req := &access.GetExecutionDataByBlockIDRequest{BlockId: blockID[:]}
+	res, err := e.execClient.GetExecutionDataByBlockID(context.Background(), req)
+	if err != nil {
+		return fmt.Errorf("could not get execution data from access node: %w", err)
+	}
+
+	execTrieUpdates := make(map[string]bool, 0)
+	// collect updates to map to compare
+	for _, gcpUpdate := range record.TrieUpdates {
+		if gcpUpdate != nil && !gcpUpdate.IsEmpty() {
+			key := gcpUpdate.String()
+			execTrieUpdates[key] = true
+			e.log.Debug().Hex("blockID", blockID[:]).Str("update", key).Msg("got update from GCP")
+		}
+	}
+	// hash search for matching updates in exec data API
+	for _, chunk := range res.GetBlockExecutionData().ChunkExecutionData {
+		convertedChunk, err := convert.MessageToChunkExecutionData(chunk, e.chain)
+		if err != nil {
+			return fmt.Errorf("unable to convert execution data chunk : %w", err)
+		}
+		tu := convertedChunk.TrieUpdate
+		lookup := tu.String()
+		if tu != nil {
```

Comment:
so there's two cases for a `nil` update:
1. There are no state changes in a given `chunk` of the exec data API (in a bitswap sense)
2. There are no trie updates for a given block, or no trie update for a transaction (in GCP).

so because there's different reasons for `nil` from the two sources, it's easier to just discard it

Reply:
Hmm that makes me a bit confused. So if we are obtaining chunks from exec API, which gets decoded to the updates and if we are downloading blocks which contain the updates, how is then even possible that all would match? I might be very wrong here but I guess chunks contain couple of blocks of updates but there would always be a border that might cut some of the blocks updates in half (some found in previous chunk and some in current) which would make comparing them always fail? Doesn't the exec API abstract that data encoding, meaning even tho bitswap syncs using chunks when you get the block execution data it gets the whole block no matter how was that transmited and synced in chunks on the network?